cloudinit/sources/DataSourceOracle.py

```diff
@@ -35,14 +36,67 @@
     'configure_secondary_nics': False,
 }
 CHASSIS_ASSET_TAG = "OracleCloud.com"
-METADATA_ROOT = "http://169.254.169.254/opc/v1/"
-METADATA_ENDPOINT = METADATA_ROOT + "instance/"
-VNIC_METADATA_URL = METADATA_ROOT + "vnics/"
 # https://docs.cloud.oracle.com/iaas/Content/Network/Troubleshoot/connectionhang.htm#Overview,
 # indicates that an MTU of 9000 is used within OCI
 MTU = 9000
 
 
+class OpcApi:
+    """Store the metadata for this instance.
+
+    Currently just stores the root_url and  metadata as dicts for each
+    of the endponits. If the requested data doesn't exist, it will be
+    fetched on first access.
+    """
+
+    def __init__(self):
+        self._root_url = None
+        self._instance_data = None
+        self._vnics_data = None
+        self.readurl = readurl
+
+    @property
+    def root_url(self):
```

Comment:
We throw away the results of the call we make here, but makes the instance_data logic a little less convoluted. If we care about that extra call, it wouldn't be hard to change.

Reply:
I think if we aren't storing this then the lazy evaluation doesn't really buy us much: in the best case with lazy evaluation, we'll make 2 calls (`instance` twice), which is the same number of calls as if we just fetched both `instance` and `vnics` unconditionally. If we _do_ need `vnics` (which will be the default behaviour at some point), then we'll make an extra call with the lazy evaluation vs. not.

Furthermore, due to it being set via `configure_secondary_nics`, we know for the entire lifecycle of the datasource instance whether or not we will need the `vnics` data or not. I think this means that we can eagerly evaluate the two fetches (if needed, in the case of `vnics`), without being any less efficient.

My suspicion is that following that path will lead to a simpler implementation (because I think we'll find that we don't need a separate class to store state for future GETs because we'll have done them all). Would you mind trying that out, then we can compare the implementations?

Reply:
I probably should have explained what motivated this change.

The old definition had these listed at the top:
```
METADATA_ROOT = "http://169.254.169.254/opc/v1/"
METADATA_ENDPOINT = METADATA_ROOT + "instance/"
VNIC_METADATA_URL = METADATA_ROOT + "vnics/"
```

Problem is, our metadata root is unknown until we make our first query. Are we querying v2 or v1? This info is needed a few times, both in and outside the class: `_get_data()` and `_get_subplatform()` inside, and `_add_network_config_from_opc_imds` and `__main__` outside (I'm ignoring `read_opc_metadata` because it's called from other parts I've mentioned). 

So I was struggling with when and where do we fetch it? I'm assuming import time is out of the question. We could keep the module level variables (METADATA_ROOT, etc), but wait to set them when we first fetch our instance data, but that's a None exception waiting to happen. 

We could move `_add_network_config_from_opc_imds` into the class and then set our metadata root / URLs (along with the instance_data) in `__init__`, but that still leaves `__main__` to deal with. Do we want to instantiate the entire DataSource in `__main__` just to get the API call? I thought it might be worthwhile to be able to make API calls without instantiating the DataSource class first, but maybe not.

For these reasons I thought it'd be simpler to move all the API related logic into it's own class so the DataSource class doesn't have to keep track of what's being fetched when.

But you're probably right that this is still a less simple implementation. I'll try the:
> We could move `_add_network_config_from_opc_imds` into the class and then set our metadata root / URLs (along with the instance_data) in `__init__`, but that still leaves `__main__` to deal with. 

option and see how that works.